packages/http-server-express/src/http-server-express.ts

```diff
@@ -0,0 +1,121 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/http-server-express
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import * as http from 'http';
+import * as https from 'https';
+import * as debugModule from 'debug';
+const debug = debugModule('loopback:http:server:express');
+
+import {
+  BaseHttpContext,
+  BaseHttpHandler,
+  HttpFactory,
+  BaseHttpEndpoint,
+  HttpServerConfig,
+  DefaultHttpEndpoint,
+} from '@loopback/http-server';
+
+import * as express from 'express';
+import {Request, Response, Application as HttpApplication} from 'express';
+
+export {
+  Request,
+  Response,
+  NextFunction,
+  Application as HttpApplication,
+} from 'express';
+
+export {
+  HttpServerConfig,
+  HttpRequestListener,
+  HttpServerLike,
+} from '@loopback/http-server';
+
+export type HttpContext = BaseHttpContext<Request, Response>;
+export type HttpHandler = BaseHttpHandler<Request, Response>;
+export type HttpEndpoint = BaseHttpEndpoint<Request, Response, HttpApplication>;
+
+function toMiddleware(handler: HttpHandler): express.RequestHandler {
+  return (request, response, next) => {
+    debug('Handling request: %s', request.originalUrl);
+    const httpCtx: HttpContext = {
+      req: request,
+      res: response,
+      request,
+      response,
+      next,
+    };
+    handler(httpCtx)
+      .then(() => {
+        debug('Finishing request: %s', request.originalUrl);
+        next();
+      })
+      .catch(err => next(err));
+  };
+}
+
+class ExpressHttpFactory
+  implements HttpFactory<Request, Response, HttpApplication> {
+  createEndpoint(config: HttpServerConfig, handler: HttpHandler) {
+    // Create an express representing the server endpoint
+    const app = express() as HttpApplication;
+    app.use(toMiddleware(handler));
+
+    let server: http.Server | https.Server;
+    if (config.protocol === 'https') {
+      server = https.createServer(config.httpsServerOptions || {}, app);
+    } else {
+      // default to http
+      server = http.createServer(app);
+    }
+    return new ExpressHttpEndpoint(config, server, app, app);
+  }
+
+  createHttpContext(
+    req: http.IncomingMessage,
+    res: http.ServerResponse,
+  ): HttpContext {
+    // Run the express middleware to parse query parameters
+    // tslint:disable-next-line:no-any
+    expressQuery()(req, res, (err: any) => {
+      if (err) throw err;
+    });
+    const request = Object.setPrototypeOf(req, expressRequest);
+    const response = Object.setPrototypeOf(res, expressResponse);
```

Comment:
See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/setPrototypeOf

> Warning: Changing the [[Prototype]] of an object is, by the nature of how modern JavaScript engines optimize property accesses, a very slow operation, in every browser and JavaScript engine. The effects on performance of altering inheritance are subtle and far-flung, and are not limited to simply the time spent in Object.setPrototypeOf(...) statement, but may extend to any code that has access to any object whose [[Prototype]] has been altered. If you care about performance you should avoid setting the [[Prototype]] of an object. Instead, create a new object with the desired [[Prototype]] using Object.create().

Can we find a different solution please, one that uses idiomatic JavaScript/TypeScript?

Also, I don't understand why is it necessary to add express prototype to the req/res objects, when `createHttpContext` is called from `toMiddleware` that already receives req/res objects from express, therefore these objects already have express additions.